src/_cffi_src/openssl/hmac.py

```diff
@@ -80,4 +79,37 @@
         return 0;
 #endif
 }
+
+HMAC_CTX *Cryptography_HMAC_CTX_new(void) {
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L && !defined(LIBRESSL_VERSION_NUMBER)
+    return HMAC_CTX_new();
+#else
+    /* This uses OPENSSL_zalloc in 1.1.0, which is malloc + memset */
+    HMAC_CTX *ctx = (HMAC_CTX *)OPENSSL_malloc(sizeof(HMAC_CTX));
+    memset(ctx, 0, sizeof(HMAC_CTX));
+    /*if (ctx)
+        if (!HMAC_CTX_reset(ctx)) {
+            HMAC_CTX_free(ctx);
+            ctx = NULL;
+        }*/
+    return ctx;
+#endif
+}
+
+
+
+void Cryptography_HMAC_CTX_free(HMAC_CTX *ctx) {
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L && !defined(LIBRESSL_VERSION_NUMBER)
+    return HMAC_CTX_free(ctx);
+#else
+    if (ctx != NULL) {
+        ctx->md = NULL;
+        ctx->key_length = 0;
+        memset(ctx->key, 0, sizeof(HMAC_MAX_MD_CBLOCK));
```

Comment:
Yeah, memsetting the memory of ctx to 0 is inclusing of setting md to NULL
and key_length to 0.

On Thu, Mar 10, 2016 at 8:35 AM, Paul Kehrer <notifications@github.com>
wrote:

> In src/_cffi_src/openssl/hmac.py
> <https://github.com/pyca/cryptography/pull/2794#discussion_r55678659>:
>
> > +            ctx = NULL;
> > +        }*/
> > +    return ctx;
> > +#endif
> > +}
> > +
> > +
> > +
> > +void Cryptography_HMAC_CTX_free(HMAC_CTX *ctx) {
> > +#if OPENSSL_VERSION_NUMBER >= 0x10100000L && !defined(LIBRESSL_VERSION_NUMBER)
> > +    return HMAC_CTX_free(ctx);
> > +#else
> > +    if (ctx != NULL) {
> > +        ctx->md = NULL;
> > +        ctx->key_length = 0;
> > +        memset(ctx->key, 0, sizeof(HMAC_MAX_MD_CBLOCK));
>
> It does? I see the memset of the whole struct, which pretty much obviates
> the need to do any of these of course. Is that what you mean?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2794/files#r55678659>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084